File: python/examples/walkthrough/write_keyset.py
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS-IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Example to showcase how to create a keyset."""
# [START tink_walkthrough_write_keyset]
from typing import TextIO

import tink
from tink import aead


def GetKmsAead(kms_kek_uri: str) -> aead.Aead:
  """Returns an AEAD primitive from a KMS Key Encryption Key URI."""
  # To obtain a primitive that uses the KMS to encrypt/decrypt we simply create
  # keyset from the appropriate template and get an AEAD primitive from it.
  template = aead.aead_key_templates.create_kms_aead_key_template(kms_kek_uri)
  kms_aead_keyset_handle = tink.new_keyset_handle(template)
  return kms_aead_keyset_handle.primitive(aead.Aead)


def WriteEncryptedKeyset(keyset_handle: tink.KeysetHandle,
                         text_io_stream: TextIO,
                         kms_kek_uri: str,
                         associated_data: bytes = b'') -> None:
  """Encrypts keyset_hanlde with a KMS and writes it to text_io_stream as JSON.

  The keyset is encrypted with a KMS using the KMS key kms_kek_uri.

  Prerequisites:
    - Register AEAD implementations of Tink.
    - Register a KMS client that can use kms_kek_uri.
    - Create a keyset and obtain a handle to it.

  Args:
    keyset_handle: Keyset to write.
    text_io_stream: I/O stream where writng the Keyset to.
    kms_kek_uri: URI of the KMS key to use to encrypt the keyset.
    associated_data: Associated data to which tie the ciphertext.

  Raises:
    tink.TinkError in case of errors.
  """
  keyset_handle.write_with_associated_data(
      tink.JsonKeysetWriter(text_io_stream), GetKmsAead(kms_kek_uri),
      associated_data)


# [END tink_walkthrough_write_keyset]

File: cc/examples/walkthrough/write_keyset.cc
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
///////////////////////////////////////////////////////////////////////////////

#include "walkthrough/write_keyset.h"

// [START tink_walkthrough_write_keyset]
#include <fstream>
#include <memory>
#include <ostream>
#include <utility>

#include "absl/status/status.h"
#include "absl/strings/string_view.h"
#include "tink/aead.h"
#include "tink/json_keyset_writer.h"
#include "tink/keyset_handle.h"
#include "tink/kms_client.h"
#include "tink/kms_clients.h"

namespace tink_walkthrough {

using ::crypto::tink::JsonKeysetWriter;
using ::crypto::tink::util::StatusOr;

// Writes a `keyset` to `output_stream` in JSON format; the keyset is encrypted
// through a KMS service using the KMS key `master_kms_key_uri`.
//
// Prerequisites for this example:
//  - Register AEAD implementations of Tink.
//  - Register a KMS client that can use `master_kms_key_uri`.
//  - Create a keyset and obtain a KeysetHandle to it.
crypto::tink::util::Status WriteEncryptedKeyset(
    const crypto::tink::KeysetHandle& keyset,
    std::unique_ptr<std::ostream> output_stream,
    absl::string_view master_kms_key_uri) {
  // Create a writer that will write the keyset to output_stream as JSON.
  StatusOr<std::unique_ptr<JsonKeysetWriter>> writer =
      JsonKeysetWriter::New(std::move(output_stream));
  if (!writer.ok()) return writer.status();
  // Get a KMS client for the given key URI.
  StatusOr<const crypto::tink::KmsClient*> kms_client =
      crypto::tink::KmsClients::Get(master_kms_key_uri);
  if (!kms_client.ok()) return kms_client.status();
  // Get an Aead primitive that uses the KMS service to encrypt/decrypt.
  StatusOr<std::unique_ptr<crypto::tink::Aead>> kms_aead =
      (*kms_client)->GetAead(master_kms_key_uri);
  if (!kms_aead.ok()) return kms_aead.status();
  return keyset.Write(writer->get(), **kms_aead);
}

}  // namespace tink_walkthrough
// [END tink_walkthrough_write_keyset]
